packages/core/core/src/TransformerRunner.js

```diff
@@ -229,6 +232,17 @@ async function finalize(asset: Asset, generate: GenerateFunc): Promise<Asset> {
   return asset;
 }
 
+async function checkPipelineCacheInvalidation(
+  pipeline: Array<Transformer>,
+  cacheEntry: CacheEntry
+): boolean {
+  return pipeline.every(
```

Comment:
There is one potential issue here, which is that an asset might switch pipelines in the middle. For example, the following Parcel config could exist:

```json
{
  "transformers": {
    "*.ts": ["@parcel/transformer-typescript"],
    "*.js": ["@parcel/transformer-babel", "@parcel/transformer-js"]
  }
}
```

In this case, a `.ts` file would first be compiled to javascript by the typescript compiler, and then run through the pipeline for `.js` files. We would need to follow all of these transformers through as well in order to ensure that `shouldInvalidateCache` is called on all of the transformers that were originally run to produce the cached asset.

The problem we have is that we don't currently store all of the intermediary steps in the cache, just the final asset at the end of all of the transformers. We also don't store which transformers ran to produce that final asset, so we currently have no way of knowing which transformers to call to check a cache entry. We may need to store this information in the cache. We should probably also store the version of the transformer plugin that generated the asset in the cache as well in order to invalidate it if you upgrade to a newer version.

Reply:
We can keep array of `appliedTransformers` in asset. Every time we run `transformer.transform` for an asset, we push `asset.appliedTransformers.push(transformer.name)`. We will save this in cache for each asset and load only these transformers while checking cache invalidation.

Reply:
I think we are going to need more information than just an array. Let's say we run a build with the following config

```
"transformers": {
  "*.ts": ["@parcel/transformer-typescript"],
  "*.js": ["@parcel/transformer-babel", "@parcel/transformer-js"]
}
```

and then on the next run we have the this config

```
"transformers": {
  "*.ts": ["@parcel/transformer-typescript"],
  "*.ts.js": ["@parcel/transformer-js"],
  "*.js": ["@parcel/transformer-babel", "@parcel/transformer-js"]
}
```

because maybe it's not necessary to run babel on typescript generated assets. In the first case, the `appliedTransformers` array would be `["@parcel/transformer-typescript", "@parcel/transformer-babel", "@parcel/transformer-js"]`. If we are just going down the line, and none of the transformers have changed, we would think the cache is valid, but it's not because it doesn't have the right applied transformers and the result may be different.

I chose to use an object where the key is the asset filepath and the value is an array of "TransformerMeta" objects which I was using to store name and version and a hash for the transformers config. So the first run would look like:

```
{
  "whatever.ts": [
    {"name": "@parcel/transformer-typescript", "version": "1.0.0", validHash: "ABC123"}
  ],
  "whatever.ts.js": [
    {"name": "@parcel/transformer-babel", "version": "1.0.0", validHash: "ABC123"},
    {"name": "@parcel/transformer-js", "version": "1.0.0", validHash: "ABC123"}
  ]
}
```

When we are checking validity on the second run, if `@parcel/transformer-typescript` has not changed we know it will produce `whatever.ts.js` again, but when we resolve the transformers for the js asset we see that the transformers have changed and the cache is invalid.

I also chose not to store this info directly on the asset because assets stick around in memory throughout the whole process and we really only need that info during the transformation phase. I suppose we could keep it on the asset as long as it is removed before we add it to the graph, but I find it simpler to just have the cache entry have meta about the transformation and the resulting assets separate. That way there's no mental hurdles of "what phases does asset have this info attached?"